doc/sphinx/language/core/sections.rst

```diff
@@ -3,57 +3,30 @@
 Section mechanism
 -----------------
 
-Sections create local contexts which can be shared across multiple definitions.
+Sections create naming scopes that permit creating section-local definitions that can
+be used by other definitions in the section.  Definitions made
+with :cmd:`Variable`, :cmd:`Hypothesis`, :cmd:`Let`, :cmd:`Let Fixpoint` and
+:cmd:`Let CoFixpoint` (or the plural variants of the first two) within sections
+are local to the section.
 
-.. example::
+In proofs done within the section, section-local definitions
+appear in the :term:`local context` of each goal in the :term:`proof state`.
```

Comment:
Yes, but in some sense this is more generic than this, it really extends the local context of any theorem or definition.
For instance, say you are in this situation:

```coq
Section foo.
Variable A : Type.
Definition id (x : A) := todo.
```

At the point of the `todo`, the local context is `(A : Type) (x : A)`, exactly as if the assumption had been in the arguments of `id` directly instead of being in the section declarations.

So "local context" is really more generic than just for proofs and it's really the whole point of sections to share (part of) a local context.

Reply:
This sentence doesn't say this is the only effect of sections--the first sentence of the subsection says "Sections create naming scopes that permit creating section-local declarations that can be used by other declarations in the section." which covers your example AFAICT.

At the moment I define that the "local context" of a goal is exactly the items that are shown above the line in the proof, some of which may be from section-level declarations.  That seems clear and precise.  How do you propose to define it precisely?  Perhaps we should use terms like section-level declaration?

Reply:
My problem is that you are making the question of "local context" specific to proofs when in fact it is not. The previous introduction did not do that restriction.

Reply:
Well, we need a term to refer to what's above the line.  It seems like a neat fit, quite precise.  But maybe it's wrong.  What behavior or properties of the "local context" are left out by making it specific to the proof?  Or is it inconsistent in some way with the implementation or how people have used the term in the past?  If I understand that then I'll be able to consider refining the definition.

Reply:
This is a perfect fit indeed. But why not describe what the local context is (a list of declarations) and then say that a goal has a conclusion (shown below the line) and a local context (shown above the line). We lose nothing compared to the current description, and yet we do not make "local context" proof-specific that way. Bodies of definitions also have a local context: it's just not displayed because the definition is not built interactively. To convince you how much it is the same, take a look at this example:

```coq
Section foo.
Variable A : Type.

(* A definition built in proof mode. *)
Definition id1 (x : A) : A.
(* Local context is (A : Type) (x : A) and conclusion (expected type) is A *)
(* Therefore I can write the following: *)
now_show A.
exact x.
Qed.

(* A definition built without the proof mode but with tactics in terms *)
Definition id2 (x : A) : A :=
(* Local context is (A : Type) (x : A) and expected type is A *)
(* Therefore I can write the following: *)
ltac:(now_show A; exact x).

(* A definition built without the proof mode and without tactics *)
Definition id3 (x : A) : A :=
(* Local context is (A : Type) (x : A) and expected type is A *)
(* Therefore I can write the following: *)
x : A.
```

In all cases, I explicitly manipulated all the variables of the local context.

Reply:
I added this sentence (_italic_):

> In proofs done within the section, section-local declarations
are included in the :term:`local context` of the initial goal of the proof.
_They are also accessible in definitions made with the :cmd:`Definition` command._

For the proof mode section and tactics, we don't use "declaration" to describe what's in the local context, we use "hypotheses", "variables" or "local definitions".  "Declaration" suggests that it's something explicitly specified by the user (and probably not changed), which doesn't fit very well for local context items such as "H : n > m" in the example.

The CiC chapter uses "declarations" and then immediately describes what constitutes a declaration.  I think the wording difference compared to proof-mode is OK, though later on, we may want to tweak "assumptions" and "definition" to be more consistent.

> A :term:`local context` is an ordered list of declarations of *variables*. The declaration of a variable :math:`x` is either an *assumption* ... or a *definition*

Is this agreeable?

Reply:
LGTM, thanks!